pkgs/development/python-modules/scipy/default.nix

```diff
@@ -17,46 +22,140 @@
+  '';
+in buildPythonPackage rec {
   pname = "scipy";
-  version = "1.10.1";
+  version = "1.11.0";
   format = "pyproject";
 
-  src = fetchPypi {
-    inherit pname version;
-    hash = "sha256-LPnfuAp7RYm6TEDOdYiYbW1c68VFfK0sKID2vC1C86U=";
+  src = fetchFromGitHub {
+    owner = "scipy";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-5mv6MKK+thoo4rvzCsN5w42x6mreIVKtTIH0JTCOu1I=";
+    # Fetch datasets that are needed for tests
+    postFetch = ''
+      # This directory must stay with this name, as it fits the environment
+      # variable XDG_CACHE_DIR set in `preConfigure`. See more details at:
+      # https://github.com/scipy/scipy/pull/18518#issuecomment-1562350648 And
+      # at: https://github.com/scipy/scipy/pull/17965#issuecomment-1560759962
+      mkdir $out/scipy-data
+    '' + (lib.concatStringsSep "\n" (builtins.map (d:
+      # From some reason, if ssl certificates are not available in this
+      # environment, see also:
+      # https://discourse.nixos.org/t/fetchfromgithub-with-a-postfetch-acts-weird/28444/7
+      "${curl} --insecure https://raw.githubusercontent.com/scipy/dataset-${d}/main/${d}.dat -o $out/scipy-data/${d}.dat"
+    ) datasets));
+    fetchSubmodules = true;
   };
-
   patches = [
-    # These tests require internet connection, currently impossible to disable
-    # them otherwise, see:
-    # https://github.com/scipy/scipy/pull/17965
-    ./disable-datasets-tests.patch
+    # Helps with cross compilation, see https://github.com/scipy/scipy/pull/18167
+    (fetchpatch {
+      url = "https://github.com/scipy/scipy/commit/dd50ac9d98dbb70625333a23e3a90e493228e3be.patch";
+      hash = "sha256-Vf6/hhwu6X5s8KWhq8bUZKtSkdVu/GtEpGtj8Olxe7s=";
+      excludes = [
+        "doc/source/dev/contributor/meson_advanced.rst"
+      ];
+    })
   ];
 
-  nativeBuildInputs = [ cython gfortran meson-python pythran pkg-config wheel ];
+  nativeBuildInputs = [ ninja cython gfortran meson meson-python pythran pkg-config wheel ];
 
   buildInputs = [
-    numpy.blas
+    blas
+    lapack
     pybind11
     pooch
+    xsimd
   ] ++ lib.optionals (pythonOlder "3.9") [
     libxcrypt
   ];
 
-  propagatedBuildInputs = [ numpy ];
+  propagatedBuildInputs = [ numpy' ];
 
   nativeCheckInputs = [ nose pytest pytest-xdist ];
 
   doCheck = !(stdenv.isx86_64 && stdenv.isDarwin);
 
   preConfigure = ''
+    # See comment near src's postFetch
+    export XDG_CACHE_HOME=$PWD
     sed -i '0,/from numpy.distutils.core/s//import setuptools;from numpy.distutils.core/' setup.py
     export NPY_NUM_BUILD_JOBS=$NIX_BUILD_CORES
   '';
 
+  # We always run what's necessary for cross compilation, see:
+  # https://github.com/mesonbuild/meson-python/pull/167
+  preBuild = ''
+    echo "Running meson setup"
+    mkdir -p build
+    meson setup \
+      --cross-file=${crossFile} \
+      -Dblas=${blasName} \
+      -Dlapack=${lapackName} \
+      build
+  '';
```

Comment:
AFAIK, the conclusion of the conversation with the upstream so far has been, that they would rather support invoking `meson setup` only once and only indirectly, by `meson-python`. I think we should try to follow their approach as far as we can. It's not an issue to move `-Dblas` and `-Dlapack` to `pipBuildFlags`, but I guess `meson-python` might be passing its own conflicting `--cross-file` in which case we might want to reach out upstream again?

Reply:
The correct approach would indeed be to pass everything through pip through meson-python. If we can't reasonable get that going we can work-around it but we should try to avoid it.

Reply:
Linking @FRidh's comment regarding the status of updating pip:

https://github.com/NixOS/nixpkgs/issues/227364#issuecomment-1612787304

Considering this pip update seems too far, I think it'd be reasonable to keep things as are - with the comment mentioning the need of a higher pip version.

Reply:
Can't say if it makes much sense, but what about introducing `pypa/build` after all? Then we can invoke meson the upstream way without waiting for pip, and we won't be triggering too many rebuilds in this PR

Reply:
> Can't say if it makes much sense, but what about introducing `pypa/build` after all?

You have a point.. Why did you drop that Idea when you commented in https://github.com/NixOS/nixpkgs/pull/230131#discussion_r1189645778 ?

> and we won't be triggering too many rebuilds in this PR

I still won't mind to add `pipBuildFlags` in anycase :).

Reply:
> You have a point.. Why did you drop that Idea when you commented in 

...because I thought we already can use pip for multiple config-settings

> I still won't mind to add pipBuildFlags in anycase :)

+1, I think there will be many more use-cases for that

Reply:
@SomeoneSerge did you manage to get pypaBuildHook working in #230131 ? I tried to pick up your work from there and I get:

```
python3.10-scipy> Executing pypaBuildPhase
python3.10-scipy> Creating a wheel...
python3.10-scipy> * Getting build dependencies for sdist...
python3.10-scipy> ERROR Missing dependencies:
python3.10-scipy>       pythran<0.14.0,>=0.12.0
python3.10-scipy>       numpy==1.21.6; python_version == "3.10" and (platform_system != "Windows" and platform_machine != "loongarch64") and platform_python_implementation != "PyPy"
```

Reply:
OK It seems that from some reason the dependencies' checks are stricter when pypa/build is used - I had to update pythran and relax the numpy dependency. I also managed to add a `setupHook` to `meson-python`, and populate `pypaBuildFlags` (and `pipBuildFlags`, although it's not relevant to us) with `mesonFlags`.